File: C++/minimum-deletions-for-at-most-k-distinct-characters.cpp
// Time:  O(n + 26)
// Space: O(n + 26)

// freq table, counting sort, greedy
class Solution {
public:
    int minDeletion(string s, int k) {
        vector<int> cnt(26);
        for (const auto& x : s) {
            ++cnt[x - 'a'];
        }
        vector<int> cnt2(ranges::max(cnt) + 1);
        for (const auto& x : cnt) {
            ++cnt2[x];
        }
        int result = 0;
        for (int i = 0, total = 26 - k; i < size(cnt2); ++i) {
            const int c = min(cnt2[i], total);
            result += i * c;
            total -= c;
            if (total == 0) {
                break;
            }
        }
        return result;
    }
};

File: Python/minimum-deletions-for-at-most-k-distinct-characters.py
# Time:  O(n + 26)
# Space: O(n + 26)

# freq table, counting sort, greedy
class Solution(object):
    def minDeletion(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        cnt = [0]*26
        for x in s:
            cnt[ord(x)-ord('a')] += 1
        cnt2 = [0]*(max(cnt)+1)
        for x in cnt:
            cnt2[x] += 1
        result = 0
        total = 26-k
        for i, x in enumerate(cnt2):
            c = min(total, x)
            result += i*c
            total -= c
            if total == 0:
                break
        return result
